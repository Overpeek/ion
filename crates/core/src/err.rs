use crate::util::{IterDisplay, Padding};
use owo_colors::OwoColorize;
use std::{
    fmt::{self, Display},
    ops::Range,
};
use thiserror::Error;

//

#[derive(Debug, Error, Clone)]
pub enum IonParseError {
    #[error("Invalid token at {row}:{col}")]
    InvalidToken {
        /// char index
        location: usize,

        /// line index
        row: usize,

        /// column index
        col: usize,
    },

    #[error("Unexpected EOF at {row}:{col}, expected one of {expected:?}")]
    UnexpectedEOF {
        /// char index
        location: usize,

        /// line index
        row: usize,

        /// column index
        col: usize,

        expected: Vec<String>,
    },

    #[error("Unexpected token at {}:{}", rows.start, cols.start)]
    UnexpectedToken {
        /// from char index to char index
        token: Range<usize>,

        /// from line index to line index
        rows: Range<usize>,

        /// from column index to colum index
        cols: Range<usize>,

        expected: Vec<String>,
    },

    #[error("Extra token")]
    ExtraToken {
        /// from char index to char index
        token: Range<usize>,

        /// from line index to line index
        rows: Range<usize>,

        /// from column index to colum index
        cols: Range<usize>,
    },

    #[error("{msg}")]
    Other { msg: String },
}

#[derive(Debug, Error, Clone, Copy)]
pub enum IonCompileError {}

#[derive(Debug, Error)]
pub enum IonError {
    #[error("Parse error: {0}")]
    Parse(IonParseError),
}

pub type IonResult<T> = Result<T, IonError>;

#[derive(Debug, Clone, Copy)]
pub struct IonPretty<'i, I> {
    inner: I,
    input: &'i str,
    file: &'i str,
    color: bool,
}

//

impl IonError {
    pub fn pretty_print<'i>(
        &'i self,
        color: bool,
        input: &'i str,
        file: &'i str,
    ) -> IonPretty<'_, &IonError> {
        IonPretty {
            inner: self,
            input,
            file,
            color,
        }
    }
}

impl IonParseError {
    pub fn pretty_print<'i>(
        &'i self,
        color: bool,
        input: &'i str,
        file: &'i str,
    ) -> IonPretty<'i, &Self> {
        IonPretty {
            inner: self,
            input,
            file,
            color,
        }
    }
}

impl fmt::Display for IonPretty<'_, &IonError> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let Self {
            inner,
            input,
            file,
            color,
        } = *self;
        match inner {
            IonError::Parse(inner) => fmt::Display::fmt(
                &IonPretty {
                    inner,
                    input,
                    file,
                    color,
                },
                f,
            ),
        }
    }
}

impl fmt::Display for IonPretty<'_, &IonParseError> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let Self {
            inner,
            input,
            file,
            color,
        } = *self;

        macro_rules! col {
            ($color:expr; $(let $i:ident: $col:ident = $s:expr;)+) => {
                $(
                let no_color = $s;
                let color = no_color.$col();
                let $i: &dyn Display = if $color { &color as _ } else { &no_color };
                )+
            };
        }

        col! {
            color;
            let error: red = "error";
            let arrow: blue = "-->";
            let bar: blue = "|";
            let hat: red = "^";
        }

        match inner {
            IonParseError::InvalidToken { row, col, .. } => {
                let (line, pad) = line_and_mag(input, *row);
                let col0 = col;
                let (row, col) = (row + 1, col + 1);
                let row_ = row;

                col! {
                    color;
                    let row: blue = row;
                };

                writeln!(f, "{error}: Invalid token")?;
                writeln!(f, "{pad}{arrow} {file}:{row_}:{col}")?;
                writeln!(f, "{pad} {bar}")?;
                writeln!(f, "{row} {bar} {line}")?;
                writeln!(f, "{pad} {bar} {:col0$}{hat}", "")?
            }
            IonParseError::UnexpectedEOF {
                row, col, expected, ..
            } => {
                let (line, pad) = line_and_mag(input, *row);
                let col0 = col;
                let (row, col) = (row + 1, col + 1);
                let row_ = row;

                col! {
                    color;
                    let row: blue = row;
                };

                writeln!(
                    f,
                    "{error}: Unexpected EOF{}",
                    IterDisplay {
                        iter: expected.iter(),
                        empty: "",
                        single: ", expected ",
                        multiple: ", expected one of ",
                        sep: ", ",
                        last_sep: " or ",
                    }
                )?;
                writeln!(f, "{pad}{arrow} {file}:{row_}:{col}")?;
                writeln!(f, "{pad} {bar}")?;
                writeln!(f, "{row} {bar} {line}")?;
                writeln!(f, "{pad} {bar} {:col0$}{hat}", "")?
            }
            IonParseError::UnexpectedToken {
                rows,
                cols,
                expected,
                ..
            } => {
                let (line, pad) = line_and_mag(input, rows.start);
                let col0 = cols.start;
                let (row, col) = (rows.start + 1, cols.start + 1);
                let row_ = row;

                col! {
                    color;
                    let row: blue = row;
                };

                writeln!(
                    f,
                    "{error}: Unexpected token{}",
                    IterDisplay {
                        iter: expected.iter(),
                        empty: "",
                        single: ", expected ",
                        multiple: ", expected one of ",
                        sep: ", ",
                        last_sep: " or ",
                    }
                )?;
                writeln!(f, "{pad}{arrow} {file}:{row_}:{col}")?;
                writeln!(f, "{pad} {bar}")?;
                writeln!(f, "{row} {bar} {line}")?;

                if !rows.is_empty() {
                    let w1 = line.len() - cols.start;
                    let w2 = cols.end;
                    writeln!(f, "{pad} {bar} {:col0$}{:^>w1$}", "", "")?;

                    let (line, pad) = line_and_mag(input, rows.end);
                    let row = rows.end + 1;

                    col! {
                        color;
                        let row: blue = row;
                    };

                    writeln!(f, "{pad} {bar}")?;
                    writeln!(f, "{row} {bar} {line}")?;
                    writeln!(f, "{pad} {bar} {:^>w2$}", "")?;
                } else {
                    let w = cols.end - cols.start;
                    writeln!(f, "{pad} {bar} {:col0$}{:^>w$}", "", "")?;
                }
            }
            IonParseError::ExtraToken { rows, cols, .. } => {
                let (line, pad) = line_and_mag(input, rows.start);
                let col0 = cols.start;
                let (row, col) = (rows.start + 1, cols.start + 1);
                let row_ = row;

                col! {
                    color;
                    let row: blue = row;
                };

                writeln!(f, "{error}: Extra token")?;
                writeln!(f, "{pad}{arrow} {file}:{row_}:{col}")?;
                writeln!(f, "{pad} {bar}")?;
                writeln!(f, "{row} {bar} {line}")?;

                if !rows.is_empty() {
                    let w1 = line.len() - cols.start;
                    let w2 = cols.end;
                    writeln!(f, "{pad} {bar} {:col0$}{:^>w1$}", "", "")?;

                    let (line, pad) = line_and_mag(input, rows.end);
                    let row = rows.end + 1;

                    col! {
                        color;
                        let row: blue = row;
                    };

                    writeln!(f, "{pad} {bar}")?;
                    writeln!(f, "{row} {bar} {line}",)?;
                    writeln!(f, "{pad} {bar} {:^>w2$}", "")?;
                } else {
                    let w = cols.end - cols.start;
                    writeln!(f, "{pad} {bar} {:col0$}{:^>w$}", "", "")?;
                }
            }
            IonParseError::Other { msg } => {
                writeln!(f, "{error}: Other error: {msg}")?;
            }
        }

        Ok(())
    }
}

//

// convert char index to line str, row and column
pub(crate) fn spot_from_location(location: usize, input: &str) -> Option<(&str, usize, usize)> {
    let mut char_counter = 0;
    input.lines().enumerate().find_map(|(row, line)| {
        let col = char_counter..char_counter + line.len() + 1;
        char_counter = col.end;
        (col.contains(&location)).then_some((line, row, location.saturating_sub(col.start)))
    })
}

fn row_to_line(input: &str, row: usize) -> &str {
    input.lines().nth(row).unwrap_or(input)
}

fn line_and_mag(input: &str, row: usize) -> (&str, Padding) {
    let line = row_to_line(input, row);
    let row_mag = if row == 0 { 1 } else { row.ilog10() + 1 };
    let pad = Padding(row_mag);

    (line, pad)
}

fn last_line_boundary_before(location: usize, input: &str) -> usize {
    input
        .chars()
        .enumerate()
        .take(location)
        .filter(|(_, ch)| *ch == '\n')
        .last()
        .map(|(i, _)| i)
        .unwrap_or(0)
}

fn first_line_boundary_after(location: usize, input: &str) -> usize {
    input
        .chars()
        .enumerate()
        .skip(location)
        .find(|(_, ch)| *ch == '\n')
        .map(|(i, _)| i)
        .unwrap_or(input.len())
}
