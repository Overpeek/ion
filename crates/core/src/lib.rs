use lalrpop_util::{lalrpop_mod, ParseError};
use serde::Serialize;

use self::{
    ast::Module,
    compiler::Compiler,
    err::{IonError, IonResult},
    ty::ResolveType,
};
use crate::err::IonParseError;

//

lalrpop_mod!(pub grammar);
pub mod ast;
pub mod compiler;
pub mod err;
pub mod prelude;
pub mod ty;
mod util;

//

/// Ion parser, interpreter and compiler
pub struct Ion {
    parser: grammar::ModuleParser,
}

//

impl Ion {
    pub fn new() -> Self {
        Self {
            parser: grammar::ModuleParser::new(),
        }
    }

    pub fn parse_str<'i>(&self, input: &'i str) -> IonResult<Module<'i>> {
        let mut ast = self.parse_str_inner(input)?;
        let mut typer = <_>::default();
        ast.type_of(&mut typer)?;
        // println!("{}", self.to_yaml(&typer));
        Ok(ast)
    }

    pub fn compile_str(&self, input: &str) -> IonResult<()> {
        let mut ast = self.parse_str(input)?;
        self.compile_ast(&mut ast)?;
        Ok(())
    }

    pub fn compile_ast(&self, ast: &mut Module) -> IonResult<()> {
        // println!("type check: {:#?}", ty::Module::new(ast));
        Compiler::compile_ast(ast, None);
        Ok(())
    }

    pub fn to_yaml(&self, ast: &impl Serialize) -> String {
        serde_yaml::to_string(&ast).unwrap()
    }

    fn parse_str_inner<'i>(&self, input: &'i str) -> IonResult<Module<'i>> {
        let mut errors = vec![];

        let res = self.parser.parse(&mut errors, input);

        let err = match res {
            Ok(module) => return Ok(module),
            Err(err) => err,
        };

        let err = match err {
            ParseError::InvalidToken { location } => {
                let (_, row, col) = err::spot_from_location(location, input)
                    .expect("Input doesn't contain the error line");

                IonParseError::InvalidToken { location, row, col }
            }
            ParseError::UnrecognizedEOF { location, expected } => {
                let (_, row, col) = err::spot_from_location(location, input)
                    .expect("Input doesn't contain the error line");

                IonParseError::UnexpectedEOF {
                    location,
                    row,
                    col,
                    expected,
                }
            }
            ParseError::UnrecognizedToken { token, expected } => {
                let (_, from_row, from_col) = err::spot_from_location(token.0, input)
                    .expect("Input doesn't contain the error line");
                let (_, to_row, to_col) = err::spot_from_location(token.2, input)
                    .expect("Input doesn't contain the error line");

                IonParseError::UnexpectedToken {
                    token: token.0..token.2,
                    rows: from_row..to_row,
                    cols: from_col..to_col,
                    expected,
                }
            }
            ParseError::ExtraToken { token } => {
                let (_, from_row, from_col) = err::spot_from_location(token.0, input)
                    .expect("Input doesn't contain the error line");
                let (_, to_row, to_col) = err::spot_from_location(token.2, input)
                    .expect("Input doesn't contain the error line");

                IonParseError::ExtraToken {
                    token: token.0..token.2,
                    rows: from_row..to_row,
                    cols: from_col..to_col,
                }
            }
            ParseError::User { error } => IonParseError::Other { msg: error },
        };

        Err(IonError::Parse(err))
    }
}

impl Default for Ion {
    fn default() -> Self {
        Self::new()
    }
}

//

#[cfg(test)]
mod tests {
    use std::process::exit;

    use crate::Ion;

    #[test]
    fn invalid_token() {
        let code = r#"x = \"#;
        let ion = Ion::new();
        let err = ion.parse_str(code).unwrap_err();

        insta::assert_display_snapshot!(err.pretty_print(false, code, "<code>"));
    }

    #[test]
    fn unexpected_eof() {
        let code = r#"x ="#;
        let ion = Ion::new();
        let err = ion.parse_str(code).unwrap_err();

        insta::assert_display_snapshot!(err.pretty_print(false, code, "<code>"));
    }

    #[test]
    fn unexpected_token_1() {
        let code = r#"x = ="#;
        let ion = Ion::new();
        let err = ion.parse_str(code).unwrap_err();

        insta::assert_display_snapshot!(err.pretty_print(false, code, "<code>"));
    }

    #[test]
    fn unexpected_token_2() {
        let code = r#""


            ""#;
        let ion = Ion::new();
        let err = ion.parse_str(code).unwrap_err();

        insta::assert_display_snapshot!(err.pretty_print(false, code, "<code>"));
    }

    #[test]
    fn correct_code() {
        let code = r#"
            x = 4;
            fn y() { x = 2; };
            y();"#;

        let ion = Ion::new();
        ion.parse_str(code).unwrap_or_else(|err| {
            eprintln!("{}", err.pretty_print(true, code, "<code>"));
            exit(-1);
        });
    }
}
