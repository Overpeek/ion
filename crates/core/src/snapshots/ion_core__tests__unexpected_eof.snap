---
source: crates/core/src/lib.rs
expression: "err.pretty_print(false, code, \"<code>\")"
---
error: Unexpected EOF, expected one of "(", "false", "fn(", "true", r#"\"[^\"]*\""#, r#"[a-zA-Z_][\\w_]*"#, r#"\\d[\\d_]*"# or r#"\\d[\\d_]+\\.\\d[\\d_]+"#
 --> <code>:1:4
  |
1 | x =
  |    ^

